src/libraries/Common/src/System/Security/Cryptography/Asn1Reader/AsnValueReader.cs

```diff
@@ -0,0 +1,232 @@
+            _span = _span.Slice(consumed);
+            return ret;
+        }
+
+        internal AsnValueReader ReadSequence(Asn1Tag? expectedTag = default)
+        {
+            AsnDecoder.ReadSequence(
+                _span,
+                _ruleSet,
+                out int contentOffset,
+                out int contentLength,
+                out int bytesConsumed,
+                expectedTag);
+
+            ReadOnlySpan<byte> content = _span.Slice(contentOffset, contentLength);
+            _span = _span.Slice(bytesConsumed);
+            return new AsnValueReader(content, _ruleSet);
+        }
+
+        internal AsnValueReader ReadSetOf(Asn1Tag? expectedTag = default)
+        {
+            AsnDecoder.ReadSetOf(
+                _span,
+                _ruleSet,
+                out int contentOffset,
+                out int contentLength,
+                out int bytesConsumed,
+                expectedTag: expectedTag);
+
+            ReadOnlySpan<byte> content = _span.Slice(contentOffset, contentLength);
+            _span = _span.Slice(bytesConsumed);
+            return new AsnValueReader(content, _ruleSet);
+        }
+
+        internal DateTimeOffset ReadUtcTime(Asn1Tag? expectedTag = default)
+        {
+            DateTimeOffset ret = AsnDecoder.ReadUtcTime(_span, _ruleSet, out int consumed, expectedTag: expectedTag);
+            _span = _span.Slice(consumed);
+            return ret;
+        }
+
+        public DateTimeOffset ReadGeneralizedTime(Asn1Tag? expectedTag = default)
+        {
+            DateTimeOffset ret = AsnDecoder.ReadGeneralizedTime(_span, _ruleSet, out int consumed, expectedTag);
+            _span = _span.Slice(consumed);
+            return ret;
+        }
+
+        public string ReadCharacterString(UniversalTagNumber encodingType, Asn1Tag? expectedTag = default)
+        {
+            string ret = AsnDecoder.ReadCharacterString(_span, _ruleSet, encodingType, out int consumed, expectedTag);
+            _span = _span.Slice(consumed);
+            return ret;
+        }
+    }
+
+    internal static class AsnWriterExtensions
+    {
+        internal static void WriteEncodedValueForCrypto(
+            this AsnWriter writer,
+            ReadOnlySpan<byte> value)
+        {
+            try
+            {
+                writer.WriteEncodedValue(value);
+            }
+            catch (ArgumentException e)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
+            }
+        }
+
+        internal static void WriteObjectIdentifierForCrypto(
+            this AsnWriter writer,
+            string value)
+        {
+            try
+            {
+                writer.WriteObjectIdentifier(value);
+            }
+            catch (ArgumentException e)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding, e);
+            }
+        }
+    }
+}
```

Comment:
Largely I've found the ArgumentException from these methods to be jarring.  I'm happy for feedback for any of:

* ArgumentException is great, leave it as-is.
* The parts that a caller can't do themselves should be AsnContentException (invalidly formed OID, invalid encoded value), but the parts they can (value smaller than 2 bytes to WriteEncodedValue, empty string to WriteObjectIdentifier) are ArgumentException.
* Like above, but without ArgumentException
* Something else

The part I don't like about it is the general principle of "an ArgumentException means the caller made a mistake [and should fix it by changing their program]." doesn't necessarily hold for these.  _Largely_ the existing code that cares about the exception shape changes is because the place that should have cared (e.g. Oid's ctor) is intentionally or historically too lax, and failures have always just come late in the game.  So maybe it's not a problem in "normal" usage.